File: mobile/app/src/main/java/com/bytecats/metanoia/bible/dao/FavoritesDao.kt
package com.bytecats.metanoia.bible.dao

import android.database.sqlite.SQLiteDatabase
import com.bytecats.metanoia.models.Favorite

class FavoritesDao(private val openDb: () -> SQLiteDatabase) {

    fun saveFavorite(strongs: String, lemma: String, definition: String) {
        val db = openDb()
        db.execSQL(
            "INSERT OR REPLACE INTO favorites (strongs, lemma, definition) VALUES (?, ?, ?)",
            arrayOf(strongs, lemma, definition)
        )
        db.close()
    }

    fun getFavorites(): List<Favorite> {
        val list = mutableListOf<Favorite>()
        val db = openDb()
        val cursor = db.rawQuery("SELECT strongs, lemma, definition FROM favorites", null)
        while (cursor.moveToNext()) list.add(Favorite(cursor.getString(0), cursor.getString(1), cursor.getString(2)))
        cursor.close(); db.close()
        return list
    }

    fun deleteFavorite(strongs: String) {
        val db = openDb()
        db.execSQL("DELETE FROM favorites WHERE strongs = ?", arrayOf(strongs))
        db.close()
    }
}

File: ios/Metanoia/Metanoia/Database/FavoritesDAO.swift
import Foundation
import GRDB

struct FavoritesDAO {
    private let db: DatabaseWriter

    init(db: DatabaseWriter) {
        self.db = db
    }

    func saveFavorite(strongs: String, lemma: String, definition: String) throws {
        try db.write { db in
            try db.execute(
                sql: "INSERT OR REPLACE INTO favorites (strongs, lemma, definition) VALUES (?, ?, ?)",
                arguments: [strongs, lemma, definition]
            )
        }
    }

    func getFavorites() throws -> [Favorite] {
        try db.read { db in
            try Favorite.fetchAll(db, sql: "SELECT strongs, lemma, definition FROM favorites")
        }
    }

    func deleteFavorite(strongs: String) throws {
        try db.write { db in
            try db.execute(
                sql: "DELETE FROM favorites WHERE strongs = ?",
                arguments: [strongs]
            )
        }
    }
}
